File: src/main/java/uk/gov/justice/digital/hmpps/nomisuserrolesapi/db/H2OmsUtils.java
package uk.gov.justice.digital.hmpps.nomisuserrolesapi.db;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public final class H2OmsUtils {

  private H2OmsUtils() {
  }

  public static void recordLogonDate(Connection conn, String username) throws SQLException {
    final String sql = "UPDATE staff_user_accounts SET last_logon_date = now() WHERE username = ?";
    try (PreparedStatement stmt = conn.prepareStatement(sql)) {
      stmt.setString(1, username);
      stmt.executeUpdate();
    }
  }

  public static void createUser(Connection conn, String username, String initialPassword, String profile) throws SQLException {
    final String createUserSql = String.format("CREATE USER %s PASSWORD ?", username);
    try (PreparedStatement createStmt = conn.prepareStatement(createUserSql)) {
      createStmt.setString(1, initialPassword);
      createStmt.executeUpdate();
    }

    final String insertSql = "INSERT INTO DBA_USERS (USERNAME, ACCOUNT_STATUS, PROFILE) values (?, ?, ?)";
    try (PreparedStatement insertStmt = conn.prepareStatement(insertSql)) {
      insertStmt.setString(1, username);
      insertStmt.setString(2, "OPEN");
      insertStmt.setString(3, profile);
      insertStmt.executeUpdate();
    }
  }

  public static void dropUser(Connection conn, String username) throws SQLException {
    final String dropUserSql = String.format("DROP USER %s", username);
    try (PreparedStatement dropStmt = conn.prepareStatement(dropUserSql)) {
      dropStmt.executeUpdate();
    }

    final String deleteSql = "DELETE FROM DBA_USERS where USERNAME = ?";
    try (PreparedStatement deleteStmt = conn.prepareStatement(deleteSql)) {
      deleteStmt.setString(1, username);
      deleteStmt.executeUpdate();
    }
  }

  public static void expirePassword(Connection conn, String username) throws SQLException {
    final String sql = "UPDATE dba_users SET account_status = ? WHERE username = ?";
    try (PreparedStatement stmt = conn.prepareStatement(sql)) {
      stmt.setString(1, "EXPIRED");
      stmt.setString(2, username);
      stmt.executeUpdate();
    }
  }

  public static void lockUser(Connection conn, String username) throws SQLException {
    final String sql = "UPDATE dba_users SET account_status = ? WHERE username = ?";
    try (PreparedStatement stmt = conn.prepareStatement(sql)) {
      stmt.setString(1, "LOCKED");
      stmt.setString(2, username);
      stmt.executeUpdate();
    }
  }

  public static void unlockUser(Connection conn, String username) throws SQLException {
    final String sql = "UPDATE dba_users SET account_status = ? WHERE username = ?";
    try (PreparedStatement stmt = conn.prepareStatement(sql)) {
      stmt.setString(1, "OPEN");
      stmt.setString(2, username);
      stmt.executeUpdate();
    }
  }

  public static void changePassword(Connection conn, String username, String password) throws SQLException {
    final String changePasswordSql = String.format("ALTER USER %s SET password ?", username);
    try (PreparedStatement stmt = conn.prepareStatement(changePasswordSql)) {
      stmt.setString(1, username);
      stmt.executeUpdate();
    }

    final String statusSql = "SELECT account_status from dba_users WHERE username = ?";
    try (PreparedStatement statusStmt = conn.prepareStatement(statusSql)) {
      statusStmt.setString(1, username);
      try (ResultSet rs = statusStmt.executeQuery()) {
        if (rs.next()) {
          final String accountStatus = rs.getString(1);
          if ("EXPIRED".equals(accountStatus)) {
            final String lockSql = "UPDATE dba_users SET account_status = ? WHERE username = ?";
            try (PreparedStatement lockStmt = conn.prepareStatement(lockSql)) {
              lockStmt.setString(1, "OPEN");
              lockStmt.setString(2, username);
              lockStmt.executeUpdate();
            }
          }
        }
      }
    }
  }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/db/H2OmsUtils.kt
package uk.gov.justice.digital.hmpps.nomisuserrolesapi.db

import java.sql.Connection
import java.sql.SQLException

class H2OmsUtils {
  companion object {
    @JvmStatic
    @Throws(SQLException::class)
    fun recordLogonDate(conn: Connection, username: String) {
      val sql = "UPDATE staff_user_accounts SET last_logon_date = now() WHERE username = ?"
      conn.prepareStatement(sql).use { stmt ->
        stmt.setString(1, username)
        stmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun createUser(conn: Connection, username: String, initialPassword: String, profile: String) {
      val createUserSql = "CREATE USER $username PASSWORD ?"
      conn.prepareStatement(createUserSql).use { createStmt ->
        createStmt.setString(1, initialPassword)
        createStmt.executeUpdate()
      }
      val insertSql = "INSERT INTO DBA_USERS (USERNAME, ACCOUNT_STATUS, PROFILE) values (?, ?, ?)"
      conn.prepareStatement(insertSql).use { insertStmt ->
        insertStmt.setString(1, username)
        insertStmt.setString(2, "OPEN")
        insertStmt.setString(3, profile)
        insertStmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun dropUser(conn: Connection, username: String) {
      val dropUserSql = "DROP USER $username"
      conn.prepareStatement(dropUserSql).use { dropStmt ->
        dropStmt.executeUpdate()
      }
      val deleteSql = "DELETE FROM DBA_USERS where USERNAME = ?"
      conn.prepareStatement(deleteSql).use { deleteStmt ->
        deleteStmt.setString(1, username)
        deleteStmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun expirePassword(conn: Connection, username: String) {
      val sql = "UPDATE dba_users SET account_status = ? WHERE username = ?"
      conn.prepareStatement(sql).use { stmt ->
        stmt.setString(1, "EXPIRED")
        stmt.setString(2, username)
        stmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun lockUser(conn: Connection, username: String) {
      val sql = "UPDATE dba_users SET account_status = ? WHERE username = ?"
      conn.prepareStatement(sql).use { stmt ->
        stmt.setString(1, "LOCKED")
        stmt.setString(2, username)
        stmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun unlockUser(conn: Connection, username: String) {
      val sql = "UPDATE dba_users SET account_status = ? WHERE username = ?"
      conn.prepareStatement(sql).use { stmt ->
        stmt.setString(1, "OPEN")
        stmt.setString(2, username)
        stmt.executeUpdate()
      }
    }

    @JvmStatic
    @Throws(SQLException::class)
    fun changePassword(conn: Connection, username: String, password: String) {
      val changePasswordSql = "ALTER USER $username SET password ?"
      conn.prepareStatement(changePasswordSql).use { stmt ->
        stmt.setString(1, username)
        stmt.executeUpdate()
      }
      val statusSql = "SELECT account_status from dba_users WHERE username = ?"
      conn.prepareStatement(statusSql).use { statusStmt ->
        statusStmt.setString(1, username)
        statusStmt.executeQuery().use { rs ->
          if (rs.next()) {
            val accountStatus = rs.getString(1)
            if ("EXPIRED" == accountStatus) {
              val lockSql = "UPDATE dba_users SET account_status = ? WHERE username = ?"
              conn.prepareStatement(lockSql).use { lockStmt ->
                lockStmt.setString(1, "OPEN")
                lockStmt.setString(2, username)
                lockStmt.executeUpdate()
              }
            }
          }
        }
      }
    }
  }
}
